File: algorithm/sort/counting/counting.js
//node counting
/**
 * counting sort
 * @param {Array} arr
 * @param {number} maxValue
 * @return {Array}
 */
function countingSort(arr, maxValue) {
    var bucket = new Array(maxValue + 1),
        sortedIndex = 0,
        arrLen = arr.length,
        bucketLen = maxValue + 1;

    for (var i = 0; i < arrLen; i++) {
        if (!bucket[arr[i]]) {
            bucket[arr[i]] = 0;
        }
        bucket[arr[i]]++;
    }

    for (var j = 0; j < bucketLen; j++) {
        while (bucket[j] > 0) {
            arr[sortedIndex++] = j;
            bucket[j]--;
        }
    }

    return arr;
}

console.log(countingSort([1, 2, 4, 3, 2, 100], 100));

File: algorithm/sort/counting/counting.go
// go run counting

package main

import "fmt"

func main()  {
	arr := countingSort([]int{1, 3, 4, 2, 6}, 6)
	fmt.Printf("result %v", arr)
}

func countingSort(arr []int, maxValue int) []int{
	bucketLen := maxValue + 1
	bucket := make([]int, bucketLen) // 初始为0的数组
	
	sortedIndex := 0
	length := len(arr)
	   
	for i := 0; i < length; i++ {
		bucket[arr[i]] ++
	}

	for j := 0; j < bucketLen; j++ {
		for bucket[j] > 0 {
			arr[sortedIndex] = j
			sortedIndex ++
			bucket[j] --
		}
     }

	return arr
}